### PR TITLE
RavenDB-18950 implementation for server-side generated Ids

### PR DESCRIPTION
**Pull request by ppekrol**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18950

### CODE REVIEW

File: src/Raven.Server/Utils/ShardHelper.cs [hunk 1]
```diff
@@ -18,6 +19,10 @@ public static class ShardHelper
     {
         public const int NumberOfBuckets = 1024 * 1024;
 
+        private static readonly Random Random = new Random();
+
+        public static int GetRandomBucket() => Random.Next() % NumberOfBuckets;
```
Comment (karmeli87) [inline src/Raven.Server/Utils/ShardHelper.cs]: This isn't thread-safe, better use `Random.Shared`

File: src/Raven.Server/Utils/ShardHelper.cs [hunk 2]
```diff
@@ -323,5 +338,19 @@ private static void ValidateBucketsMapping(DatabaseRecord record)
                 lastShard = current.ShardNumber;
             }
         }
+
+        public static string GenerateStickyId(string id, int bucket, char identityPartsSeparator)
+        {
+            Debug.Assert(id[^1] == identityPartsSeparator, "id[^1] == identityPartsSeparator");
+
+            return $"{id}${bucket}{SuffixIdTerminator}";
```
Comment (karmeli87) [inline src/Raven.Server/Utils/ShardHelper.cs]: so the hash will be generated based on `{bucket}{SuffixIdTerminator}` ?

File: src/Raven.Server/Documents/DocumentPutAction.cs [hunk 3]
```diff
@@ -429,43 +429,73 @@ private string BuildDocumentId(string id, long newEtag, out bool knownNewId)
...
-                    string value = new string('0', id.Length + 1 + 19 + nodeTag.Length);
-                    fixed (char* valuePtr = value)
+                    var isIdentityPartsSeparator = lastChar == _documentDatabase.IdentityPartsSeparator;
+
+                    if (isIdentityPartsSeparator == false && lastChar == ShardHelper.SuffixIdTerminator)
                     {
-                        char* valueTagPtr = valuePtr + value.Length - nodeTag.Length;
-                        for (int j = 0; j < nodeTag.Length; j++)
-                            valueTagPtr[j] = nodeTag[j];
+                        idSuffixLength = id.Length;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/DocumentPutAction.cs]: do we want to throw/validate here that we are on the proper shard?

File: src/Raven.Server/Utils/ShardHelper.cs (hunk 2 above)
Comment (karmeli87) [reply src/Raven.Server/Utils/ShardHelper.cs]: Sorry it is a bit hard for me to follow the end result here
From the orchestrator we generate something like `Users/$123*` and sending this to the shard which is the result of hashing only the `123`, right?
But on the shard itself we end up with what kind of id?
if it is `users/00000565-A$123`, then we okay.

**Review by karmeli87 — COMMENTED** [review]
The idea is great, but if I understand correctly the implementation feels a bit too hacky to me..
Can we maybe pass some extra info on the request to the shard and propagate that to the `BuildDocumentId` method?

**Comment by karmeli87** [comment]
Also I know that `BuildDocumentId` is inlined but we can consider of moving it into the `DocumentDatabase` and override in the `ShardedDocumentDatabase`

File: src/Raven.Server/Documents/DocumentPutAction.cs [hunk 4]
```diff
@@ -429,43 +429,73 @@ private string BuildDocumentId(string id, long newEtag, out bool knownNewId)
...
+                    if (isIdentityPartsSeparator)
+                    {
+                        string nodeTag = _documentDatabase.ServerStore.NodeTag;
+
+                        // PERF: we are creating an string and mutating it for performance reasons.
+                        //       while nasty this shouldn't have any side effects because value didn't
+                        //       escape yet the function, so while not pretty it works (and it's safe).
+                        //       
+                        int valueLength = idLength + 1 + 19 + nodeTag.Length + idSuffixLength;
+                        string value = new('0', valueLength);
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentPutAction.cs:463]: Shouldn't we use `string.Create` here?

File: src/Raven.Server/Documents/DocumentPutAction.cs [hunk 5]
```diff
@@ -429,43 +429,73 @@ private string BuildDocumentId(string id, long newEtag, out bool knownNewId)
...
+                var lastChar = id[^1];
                 if (lastChar == '|')
                 {
                     ThrowInvalidDocumentId(id);
                 }
 
-                if (lastChar == _documentDatabase.IdentityPartsSeparator)
-                {                    
-                    string nodeTag = _documentDatabase.ServerStore.NodeTag;
+                fixed (char* idPtr = id)
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentPutAction.cs:446]: why do we need this `fixed` ?

File: src/Raven.Server/Utils/ShardHelper.cs [hunk 6]
```diff
@@ -323,5 +338,19 @@ private static void ValidateBucketsMapping(DatabaseRecord record)
...
+        }
+
+        public static unsafe void ExtractStickyId(ref char* buffer, ref int size)
+        {
+            AdjustAfterSeparator('$', ref buffer, ref size);
+        }
+
+        public const char SuffixIdTerminator = '*';
```
Comment (ayende) [inline src/Raven.Server/Utils/ShardHelper.cs]: That is a currently a valid document id char, no?
What about doing something like `users/$12312/` And then we can detect the `/$` previously and react

File: src/Raven.Server/Documents/Sharding/Handlers/Batches/ShardedBatchCommand.cs [hunk 7]
```diff
@@ -79,18 +79,28 @@ public IEnumerator<SingleShardedCommand> GetEnumerator()
                 continue;
             }
 
-            var shard = _databaseContext.GetShardNumber(_context, cmd.Id);
-            var commandStream = bufferedCommand.CommandStream;
-            var stream = cmd.Type == CommandType.AttachmentPUT ? AttachmentStreams[streamPosition++] : null;
+            int shardNumber;
+            if (bufferedCommand.IsServerSideIdentity)
+            {
+                var bucket = ShardHelper.GetRandomBucket(); // we do not care about bucket here, just need a random number with max value
```
Comment (ayende) [inline src/Raven.Server/Documents/Sharding/Handlers/Batches/ShardedBatchCommand.cs]: That isn't actually a bucket id. It is just something that we hash.
We are probably better generating a alphanumeric value here, instead.
So generate a random value, then use the Base26 encoding to get a random looking value.

A numeric value may lead us to assume this _is_ the bucket.

Comment (karmeli87) [reply src/Raven.Server/Documents/Sharding/Handlers/Batches/ShardedBatchCommand.cs]: I agree. 
If my math is correct we will need only 5 chars in Base26. (given that the hash distribution is uniform)

**Pull request merged**